### ISSUE TITLE
About demos 

### ISSUE THREAD

**Issue by JoyHuYY1412** [SHI-Labs/Semi-Supervised-Transfer-Learning#1]
Would you mind provide more demos for this repo? BTW, it seems it cannot support CIAFR10 currently. Thank you so much!

**Comment by Walleclipse** [comment]
Hi, 
This code can be used for CUB-200-2011, MIT Indoor-67, and MURA datasets. 
1) Download the [CUB-200-2011 dataset](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html).
2) Pretrain the Resnet-50 model on [Imagenet](http://image-net.org/download-images) or download the Imagenet pretrained models from [torchvision.models](https://pytorch.org/vision/0.8/models.html).
PS: The output of our model is slightly different from the original Resnet-50 model, please check the code.
3) In [parser.py](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/parser.py), `lambda_kd` represents the regularization weight factors for AKC, and  `lambda_mmd` represents the regularization weight factors for ARC. 
e.g: Train the models with ARC:
`
python main.py --lambda_mmd 50  --lambda_kd 0 
`
Train the models with AKC:
`
python main.py --lambda_mmd 0  --lambda_kd 1 
`

CIFAR-10 training has quite different hyperparameters and models, please check the setting of the CIFAR-10 on the paper of [FixMatch](https://github.com/google-research/fixmatch) and [Realistic-Evaluation-SSL](https://github.com/brain-research/realistic-ssl-evaluation) or [SSL-code](https://github.com/perrying/pytorch-consistency-regularization). You need to pre-train a Wide ResNet-28-2 model on [ImageNet 32×32](https://patrykchrabaszcz.github.io/Imagenet32/), then transfer it to the CIFAR-10 dataset.

As CIFAR-10 is well studied in SSL works, it's worthy to investigate more difficult real-world images. In this work,  we mainly consider the popular transfer learning benchmark datasets.

**Comment by JoyHuYY1412** [comment]
> Hi,
> This code can be used for CUB-200-2011, MIT Indoor-67, and MURA datasets.
> 
> 1. Download the [CUB-200-2011 dataset](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html).
> 2. Pretrain the Resnet-50 model on [Imagenet](http://image-net.org/download-images) or download the Imagenet pretrained models from [torchvision.models](https://pytorch.org/vision/0.8/models.html).
>    PS: The output of our model is slightly different from the original Resnet-50 model, please check the code.
> 3. In [parser.py](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/parser.py), `lambda_kd` represents the regularization weight factors for AKC, and  `lambda_mmd` represents the regularization weight factors for ARC.
>    e.g: Train the models with ARC:
>    `python main.py --lambda_mmd 50  --lambda_kd 0 `
>    Train the models with AKC:
>    `python main.py --lambda_mmd 0  --lambda_kd 1 `
> 
> CIFAR-10 training has quite different hyperparameters and models, please check the setting of the CIFAR-10 on the paper of [FixMatch](https://github.com/google-research/fixmatch) and [Realistic-Evaluation-SSL](https://github.com/brain-research/realistic-ssl-evaluation) or [SSL-code](https://github.com/perrying/pytorch-consistency-regularization). You need to pre-train a Wide ResNet-28-2 model on [ImageNet 32×32](https://patrykchrabaszcz.github.io/Imagenet32/), then transfer it to the CIFAR-10 dataset.
> 
> As CIFAR-10 is well studied in SSL works, it's worthy to investigate more difficult real-world images. In this work, we mainly consider the popular transfer learning benchmark datasets.

Thank you for your reply! I will try it!

**Comment by JoyHuYY1412** [comment]
> Hi,
> This code can be used for CUB-200-2011, MIT Indoor-67, and MURA datasets.
> 
> 1. Download the [CUB-200-2011 dataset](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html).
> 2. Pretrain the Resnet-50 model on [Imagenet](http://image-net.org/download-images) or download the Imagenet pretrained models from [torchvision.models](https://pytorch.org/vision/0.8/models.html).
>    PS: The output of our model is slightly different from the original Resnet-50 model, please check the code.
> 3. In [parser.py](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/parser.py), `lambda_kd` represents the regularization weight factors for AKC, and  `lambda_mmd` represents the regularization weight factors for ARC.
>    e.g: Train the models with ARC:
>    `python main.py --lambda_mmd 50  --lambda_kd 0 `
>    Train the models with AKC:
>    `python main.py --lambda_mmd 0  --lambda_kd 1 `
> 
> CIFAR-10 training has quite different hyperparameters and models, please check the setting of the CIFAR-10 on the paper of [FixMatch](https://github.com/google-research/fixmatch) and [Realistic-Evaluation-SSL](https://github.com/brain-research/realistic-ssl-evaluation) or [SSL-code](https://github.com/perrying/pytorch-consistency-regularization). You need to pre-train a Wide ResNet-28-2 model on [ImageNet 32×32](https://patrykchrabaszcz.github.io/Imagenet32/), then transfer it to the CIFAR-10 dataset.
> 
> As CIFAR-10 is well studied in SSL works, it's worthy to investigate more difficult real-world images. In this work, we mainly consider the popular transfer learning benchmark datasets.

HI, I ran the code using the pre-trained ResNet model (https://download.pytorch.org/models/resnet50-19c8e357.pth) with ARC under CUB@400.  But I get following results:

`mean test acc. over last 1 checkpoints: 0.38246461857093544
median test acc. over last 1 checkpoints: 0.38246461857093544
mean test acc. over last 10 checkpoints: 0.3827407663387803
median test acc. over last 10 checkpoints: 0.3825509147393856
mean test acc. over last 20 checkpoints: 0.38154124960401037
median test acc. over last 20 checkpoints: 0.3822920262546098
mean test acc. over last 50 checkpoints: 0.3746082154277443
median test acc. over last 50 checkpoints: 0.3761649982997949
`

And it seems the accuracy is not improved too much during training.  Could you please check it and give me some advice? Thank you so much!

**Comment by Walleclipse** [comment]
> 
> HI, I ran the code using the pre-trained ResNet model (https://download.pytorch.org/models/resnet50-19c8e357.pth) with ARC under CUB@400. But I get following results:
> 
> `mean test acc. over last 1 checkpoints: 0.38246461857093544 median test acc. over last 1 checkpoints: 0.38246461857093544 mean test acc. over last 10 checkpoints: 0.3827407663387803 median test acc. over last 10 checkpoints: 0.3825509147393856 mean test acc. over last 20 checkpoints: 0.38154124960401037 median test acc. over last 20 checkpoints: 0.3822920262546098 mean test acc. over last 50 checkpoints: 0.3746082154277443 median test acc. over last 50 checkpoints: 0.3761649982997949 `
> 
> And it seems the accuracy is not improved too much during training. Could you please check it and give me some advice? Thank you so much!

Hi,
I added some description about the code [here](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/tree/main/code), please check it. 
1. I slightly modified some hyperparameters, including `epochs`. In general, the more training epochs lead the better performance. Please check the [parser.py](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/code/parser.py).
2. I downloaded the pre-trained model from [here](https://download.pytorch.org/models/resnet50-19c8e357.pth). You can try the same model as me. 
3. You can try modifying some of the hyperparameters. The important hyperparameters including `lr`,`epochs`,`lambda_mmd`,`mmd_threshold`,`mmd_feat_table_l`, `mmd_feat_table_u`. 
4. Please plot the training curve according to [plot_loss.ipynb](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/code/plot_loss.ipynb). If you get a similar evaluation accuracy curve as below, the results might be right. (Of course, the exact value of the performance in an experiment is affected by some randomness of the running system.)
![acc_curve](https://user-images.githubusercontent.com/37198574/110798985-a68ecd00-82b5-11eb-86e9-2eb12ae1797d.png)

**Comment by JoyHuYY1412** [comment]
> > HI, I ran the code using the pre-trained ResNet model (https://download.pytorch.org/models/resnet50-19c8e357.pth) with ARC under CUB@400. But I get following results:
> > `mean test acc. over last 1 checkpoints: 0.38246461857093544 median test acc. over last 1 checkpoints: 0.38246461857093544 mean test acc. over last 10 checkpoints: 0.3827407663387803 median test acc. over last 10 checkpoints: 0.3825509147393856 mean test acc. over last 20 checkpoints: 0.38154124960401037 median test acc. over last 20 checkpoints: 0.3822920262546098 mean test acc. over last 50 checkpoints: 0.3746082154277443 median test acc. over last 50 checkpoints: 0.3761649982997949 `
> > And it seems the accuracy is not improved too much during training. Could you please check it and give me some advice? Thank you so much!
> 
> Hi,
> I added some description about the code [here](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/tree/main/code), please check it.
> 
> 1. I slightly modified some hyperparameters, including `epochs`. In general, the more training epochs lead the better performance. Please check the [parser.py](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/code/parser.py).
> 2. I downloaded the pre-trained model from [here](https://download.pytorch.org/models/resnet50-19c8e357.pth). You can try the same model as me.
> 3. You can try modifying some of the hyperparameters. The important hyperparameters including `lr`,`epochs`,`lambda_mmd`,`mmd_threshold`,`mmd_feat_table_l`, `mmd_feat_table_u`.
> 4. Please plot the training curve according to [plot_loss.ipynb](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/code/plot_loss.ipynb). If you get a similar evaluation accuracy curve as below, the results might be right. (Of course, the exact value of the performance in an experiment is affected by some randomness of the running system.)
>    ![acc_curve](https://user-images.githubusercontent.com/37198574/110798985-a68ecd00-82b5-11eb-86e9-2eb12ae1797d.png)

Thank you for your patience! I run the modified code, however, the result of ARC is comparable with AKC (~39%). Could you please check it? I will also try to find the reason.

**Comment by Walleclipse** [comment]
> Thank you for your patience! I run the modified code, however, the result of ARC is comparable with AKC (~39%). Could you please check it? I will also try to find the reason.

Hi, I downloaded the repo and re-ran the code. I got similar results (ARC ~ 41%) with the following command.
```
pretrain_path="/ssd3/abu/code/semi_transfer/consistency_regularization/ssl_lib/models/ckpt" # folder path of your pretrained model
data_root="/ssd3/abu/code/semi_transfer/consistency_regularization/data" # folder path of your data
dataset=cub200
num_labels=400
arc=50 # adaptive representation consistency (semi-supervised)
akc=0 # adaptive knowledge consistency (transfer)

CUDA_VISIBLE_DEVICES=0,1 nohup  python -u main.py \
--data_root $data_root --dataset $dataset --num_labels $num_labels --pretrained_weight_path $pretrain_path  \
--lambda_mmd $arc --lambda_kd $akc --epochs 100 \
>logs_${dataset}_${num_labels}_${arc}_${akc}.out 2>&1 &
```

1. Please compare your training log with my  [logs_cub200_400_50_0.txt](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/files/6168909/logs_cub200_400_50_0.txt), especially check the hyperparameters in the training log.

2. Please plot the training curve according to [plot_loss.ipynb](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/code/plot_loss.ipynb) and compare it with my training curve.

**Comment by JoyHuYY1412** [comment]
> > Thank you for your patience! I run the modified code, however, the result of ARC is comparable with AKC (~39%). Could you please check it? I will also try to find the reason.
> 
> Hi, I downloaded the repo and re-ran the code. I got similar results (ARC ~ 41%) with the following command.
> 
> ```
> pretrain_path="/ssd3/abu/code/semi_transfer/consistency_regularization/ssl_lib/models/ckpt" # folder path of your pretrained model
> data_root="/ssd3/abu/code/semi_transfer/consistency_regularization/data" # folder path of your data
> dataset=cub200
> num_labels=400
> arc=50 # adaptive representation consistency (semi-supervised)
> akc=0 # adaptive knowledge consistency (transfer)
> 
> CUDA_VISIBLE_DEVICES=0,1 nohup  python -u main.py \
> --data_root $data_root --dataset $dataset --num_labels $num_labels --pretrained_weight_path $pretrain_path  \
> --lambda_mmd $arc --lambda_kd $akc --epochs 100 \
> >logs_${dataset}_${num_labels}_${arc}_${akc}.out 2>&1 &
> ```
> 
> 1. Please compare your training log with my  [logs_cub200_400_50_0.txt](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/files/6168909/logs_cub200_400_50_0.txt), especially check the hyperparameters in the training log.
> 2. Please plot the training curve according to [plot_loss.ipynb](https://github.com/SHI-Labs/Semi-Supervised-Transfer-Learning/blob/main/code/plot_loss.ipynb) and compare it with my training curve.

Hi, I use your command and get the same result. Thank you so much! The interesting thing is that when I do not set the available devices, here I would use 8 Tesla P100 in default, the performance is 2 percent lower. 

I am confused because I think the data-parallel function would not influence the final result.   Maybe because of the batch-norm.

**Comment by Walleclipse** [comment]
> 
> Hi, I use your command and get the same result. Thank you so much! The interesting thing is that when I do not set the available devices, here I would use 8 Tesla P100 in default, the performance is 2 percent lower.
> 
> I am confused because I think the data-parallel function would not influence the final result. Maybe because of the batch-norm.

Thank you for your experiments. I didn't notice the problem before.  As you mentioned, the performance-degeneration in multi-GPU maybe because of the batch-norm update of labeled and unlabeled examples in different GPU devices. The related issue can be found in  [fixmatch-issues](https://github.com/google-research/fixmatch/issues/20).

**Comment by JoyHuYY1412** [comment]
> > Hi, I use your command and get the same result. Thank you so much! The interesting thing is that when I do not set the available devices, here I would use 8 Tesla P100 in default, the performance is 2 percent lower.
> > I am confused because I think the data-parallel function would not influence the final result. Maybe because of the batch-norm.
> 
> Thank you for your experiments. I didn't notice the problem before. As you mentioned, the performance-degeneration in multi-GPU maybe because of the batch-norm update of labeled and unlabeled examples in different GPU devices. The related issue can be found in [fixmatch-issues](https://github.com/google-research/fixmatch/issues/20).

Thank you for your patience! I am new to semi-supervise learning and your advice really helps!
